src/expr/src/scalar/func/impls/timestamp.rs

```diff
@@ -61,38 +61,182 @@ sqlfunc!(
+        f.write_str("timestamp_to_timestamp_with_time_zone")
+    }
+}
+
+#[derive(
+    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct CastTimestampToTimestamp(pub Option<TimestampPrecision>);
+
+impl<'a> EagerUnaryFunc<'a> for CastTimestampToTimestamp {
+    type Input = CheckedTimestamp<NaiveDateTime>;
+    type Output = Result<CheckedTimestamp<NaiveDateTime>, EvalError>;
+
+    fn call(
+        &self,
+        mut a: CheckedTimestamp<NaiveDateTime>,
+    ) -> Result<CheckedTimestamp<NaiveDateTime>, EvalError> {
+        a.round_to_precision(self.0);
+        Ok(a)
+    }
+
+    fn output_type(&self, input: ColumnType) -> ColumnType {
+        ScalarType::Timestamp { precision: self.0 }.nullable(input.nullable)
+    }
+
+    fn preserves_uniqueness(&self) -> bool {
+        // Only preserves uniqueness if the output type is as precise as the
+        // input type, but we're not provided in the input type here.
+        false
+    }
+
+    fn inverse(&self) -> Option<crate::UnaryFunc> {
+        None
+    }
+
+    fn is_monotone(&self) -> bool {
+        true
+    }
+}
+
+impl fmt::Display for CastTimestampToTimestamp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("timestamp_to_timestamp")
+    }
+}
+
+#[derive(
+    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct CastTimestampTzToTimestamp(pub Option<TimestampPrecision>);
+
+impl<'a> EagerUnaryFunc<'a> for CastTimestampTzToTimestamp {
+    type Input = CheckedTimestamp<DateTime<Utc>>;
+    type Output = Result<CheckedTimestamp<NaiveDateTime>, EvalError>;
+
+    fn call(
+        &self,
         a: CheckedTimestamp<DateTime<Utc>>,
     ) -> Result<CheckedTimestamp<NaiveDateTime>, EvalError> {
-        a.naive_utc().try_into().err_into()
+        let mut out = CheckedTimestamp::try_from(a.naive_utc())?;
+        out.round_to_precision(self.0);
+        Ok(out)
     }
-);
+
+    fn output_type(&self, input: ColumnType) -> ColumnType {
+        ScalarType::Timestamp { precision: self.0 }.nullable(input.nullable)
+    }
+
+    fn preserves_uniqueness(&self) -> bool {
+        // Only preserves uniqueness if the output type is as precise as the
+        // input type, but we're not provided in the input type here.
+        false
+    }
+
+    fn inverse(&self) -> Option<crate::UnaryFunc> {
+        to_unary!(super::CastTimestampToTimestampTz(self.0))
+    }
+
+    fn is_monotone(&self) -> bool {
+        true
+    }
+}
+
+impl fmt::Display for CastTimestampTzToTimestamp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("timestamp_with_time_zone_to_timestamp")
+    }
+}
+
+#[derive(
+    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct CastTimestampTzToTimestampTz(pub Option<TimestampPrecision>);
```

Comment:
nit: I prefer for these functions to have a more semantically meaningful name, e.g. `RescaleTimestampTzPrecision`--or we should rename `RescaleNumeric` to `CastNumericToNumeric`. Comment also applies to `CastTimestampToTimestamp`.

Reply:
Ah yeah that's a great point, Sean. I'd go `AdjustTimestampTzPrecision` since "rescale" is pretty specifically tied to adjusting a decimal scale. Too bad "reprecision" isn't a term. I think ideally the names would be `AdjustNumericScale` and `AdjustTimestampTzPrecision` for perfect symmetry.

Reply:
Agreed, a semantically meaningful name would be better here. Let me do the renames as a follow-up.